### PR TITLE
Added Exercise 3

### PR DESCRIPTION
**Pull request by Saurabhrkp**
Added all exercise 3.1, 3.2, 3.3 & 3.4

### CODE REVIEW

File: Week-3/exercise_3_1/index.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+//Given reducer method:
+const add = (a, b = 0) => a + b;
+
+const memoize = (func) => {
+  // a cache of results
+  const results = {};
+  return (...args) => {
+    const argsKey = JSON.stringify(args);
+    if (!results[argsKey]) {
```
Comment (rahulm2) [inline Week-3/exercise_3_1/index.js:9]: @Saurabhrkp question will this search for property in prototype chain?

File: Week-3/exercise_3_2/index.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  console.log(`${this.brand} - ${this.model} owned by ${ownerName} was created on ${date}`);
+};
+
+carCreatedOn.apply(car, ['Sameer', '2019-01-01']);
+
+const carDetails = function (type) {
+  console.log(`${this.brand} - ${this.model} is an ${type} car`);
+};
+
+carDetails.call(car, 'Electric');
```
Comment (rahulm2) [inline Week-3/exercise_3_2/index.js:24]: add new line

Comment (rahulm2) [reply Week-3/exercise_3_2/index.js:24]: @Saurabhrkp can u also explain pros and cons of each

File: Week-3/exercise_3_3/index.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+const [increment, log] = createIncrement();
+
+increment();
+increment();
+increment();
+log();
+
+// What is logged?
+// It will log: Count is 0
```
Comment (rahulm2) [inline Week-3/exercise_3_3/index.js:21]: Add explanation why it will be 0 and also add new line

**Comment by rahulm2** [comment]
@Saurabhrkp can u also use a Pull request template

File: Week-3/exercise_3_1/index.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+//Given reducer method:
+const add = (a, b = 0) => a + b;
```
Comment (rahulm2) [inline Week-3/exercise_3_1/index.js]: @Saurabhrkp the assignment asks for n number of args. This fn only processes 2

Comment (Saurabhrkp) [reply Week-3/exercise_3_1/index.js]: Okay, making add function to get n number of args

File: Week-3/exercise_3_1/index.js (hunk 1 above)
Comment (Saurabhrkp) [reply Week-3/exercise_3_1/index.js:9]: is it required to seacrh

File: Week-3/exercise_3_1/index.js (hunk 4 above)
Comment (Saurabhrkp) [reply Week-3/exercise_3_1/index.js]: done

File: Week-3/exercise_3_1/index.js (hunk 1 above)
Comment (Saurabhrkp) [reply Week-3/exercise_3_1/index.js:9]: yes it will search after it complete searching in properties

File: Week-3/exercise_3_2/index.js (hunk 2 above)
Comment (Saurabhrkp) [reply Week-3/exercise_3_2/index.js:24]: In JavaScript, bind, call, and apply are all methods that can be used to change the this context of a function, and allow you to call a function with a specific this value. Here are some pros and cons of each method:

bind:

Pros:
It returns a new function with a fixed this context, which can be useful for creating functions with a specific this value that can be called later.
It allows you to pre-set some of the arguments for the function, so they don't need to be passed in when the function is called.
Cons:
It doesn't call the function immediately, so it's less useful for functions that need to be called right away.
call:

Pros:
It allows you to change the this context of a function and call it immediately, which can be useful for functions that need to be called right away.
It allows you to pass in arguments individually, which can be useful for functions with a variable number of arguments.
Cons:
It doesn't return a new function with a fixed this context, so you'll need to call it with the correct this value every time you want to use it.
apply:

Pros:
Like call, it allows you to change the this context of a function and call it immediately.
It allows you to pass in arguments as an array, which can be useful for functions with a variable number of arguments.
Cons:
Like call, it doesn't return a new function with a fixed this context, so you'll need to call it with the correct this value every time you want to use it.
In summary, bind is useful for creating reusable functions with a fixed this context, call and apply are useful for changing the this context of a function and calling it immediately. It depends on the use case and requirement of the function whether to use bind, call or apply.

File: Week-3/exercise_3_3/index.js (hunk 3 above)
Comment (Saurabhrkp) [reply Week-3/exercise_3_3/index.js:21]: Because message is already defined on initizalition  let message = `Count is ${count}`;
